### PR TITLE
[WIP] Convert .md links to .html

### PR DESCRIPTION
**Pull request by wxwxwxwx9**
**What is the purpose of this pull request? (put "X" next to an item, remove the rest)**
• [x] Enhancement to an existing feature

This PR addresses issue #1227. 

**What is the rationale for this request?**
Refer to issue #1227 thread, it is well-elaborated there. 

**What changes did you make? (Give an overview)**
I implemented a static method under ComponentParser to convert all the existing files with **.md** extension to **.html** extension (if the user did not specify the 'no-convert' attribute in their Markdown code).

The method is called in the render() method of ComponentParser. 

**Is there anything you'd like reviewers to focus on?**
Code quality and edge cases (where I convert files with **.md** extensions which I should not be converting)

**Testing instructions:**
```npm run test```

**Proposed commit message: (wrap lines at 72 characters)**
Convert .md links to .html during render

### CODE REVIEW

File: packages/core/src/parsers/ComponentParser.js [hunk 1]
```diff
@@ -614,6 +614,20 @@ class ComponentParser {
     return node;
   }
 
+  static convertMdExtToHtmlExt(renderedContent) {
+    const $ = cheerio.load(renderedContent);
```
Comment (ang-zeyu) [inline packages/core/src/parsers/ComponentParser.js:618]: 👀 

`cheerio.load` is really convenient but expensive (it parses the entire page one more time);

you can look at the other functions for how to manipulate the pseudo dom (which is also still usable with cheerio - so we don't sacrifice code quality for more complex operations) (but still you shouldn't need cheerio for the simple conversion here)

wondering if we should move this function to `linkProcessor` since it deals with links as well

File: packages/core/src/parsers/ComponentParser.js [hunk 2]
```diff
@@ -642,7 +656,8 @@ class ComponentParser {
       const parser = new htmlparser.Parser(handler);
       const fileExt = utils.getExt(file);
       if (utils.isMarkdownFileExt(fileExt)) {
-        const renderedContent = md.render(content);
+        let renderedContent = md.render(content);
+        renderedContent = ComponentParser.convertMdExtToHtmlExt(renderedContent);
```
Comment (ang-zeyu) [inline packages/core/src/parsers/ComponentParser.js:660]: you want to modify `_parse` so `html` files get the auto conversion too

**Review by ang-zeyu — COMMENTED** [review]
quick look:

File: packages/core/src/parsers/ComponentParser.js (hunk 2 above)
Comment (wxwxwxwx9) [reply packages/core/src/parsers/ComponentParser.js]: Shifted the invocation into ```_parse``` function!

File: packages/core/src/parsers/ComponentParser.js (hunk 1 above)
Comment (wxwxwxwx9) [reply packages/core/src/parsers/ComponentParser.js]: Yup, thanks for highlighting the expensive nature of ```cheerio.load```. 

I have modified the function to modify nodes directly as we are traversing through them in the ```_parse``` function. 

I have also shifted the function into `linkProcessor`!

File: packages/core/src/parsers/linkProcessor.js [hunk 3]
```diff
@@ -45,6 +45,19 @@ function convertRelativeLinks(node, cwf, rootPath, baseUrl) {
   node.attribs[linkAttribName] = path.posix.join(baseUrl || '/', resourcePathFromRoot);
 }
 
+function convertMdExtToHtmlExt(node) {
+  if (node.attribs && node.attribs.href) {
+    const { href } = node.attribs;
+    const hasMdExtension = href && href.slice(-3) === '.md';
```
Comment (ang-zeyu) [inline packages/core/src/parsers/linkProcessor.js:51]: first `href &&` can be removed since you checked it above already

File: packages/core/src/parsers/linkProcessor.js [hunk 4]
```diff
@@ -45,6 +45,19 @@ function convertRelativeLinks(node, cwf, rootPath, baseUrl) {
   node.attribs[linkAttribName] = path.posix.join(baseUrl || '/', resourcePathFromRoot);
 }
 
+function convertMdExtToHtmlExt(node) {
+  if (node.attribs && node.attribs.href) {
```
Comment (ang-zeyu) [inline packages/core/src/parsers/linkProcessor.js:49]: should we limit this to anchor tags as well? `node.name === 'a'`, since some other tags may use `href` for different purposes and have unintended side effects

**Review by ang-zeyu — COMMENTED** [review]
thanks, some more minor nits, then we can merge this in : )

**Comment by ang-zeyu** [comment]
will need to update the ug section too https://markbind.org/userGuide/formattingContents.html#links

File: packages/core/src/parsers/linkProcessor.js (hunk 4 above)
Comment (wxwxwxwx9) [reply packages/core/src/parsers/linkProcessor.js:49]: agreed! thanks for bringing this up!

File: packages/core/src/parsers/linkProcessor.js (hunk 3 above)
Comment (wxwxwxwx9) [reply packages/core/src/parsers/linkProcessor.js:51]: ah yeah, thanks for highlighting this, I overlooked it!

File: docs/userGuide/syntax/links.mbdf [hunk 5]
```diff
@@ -30,6 +30,27 @@ MarkBind home is at [here][1].
...
+<div class="indented">
+
+Do indicate markdown files with its original extension (**.md**) as it will automatically be converted to a html extension (**.html**) when the site is generated. 
+
+{{ icon_example }}
+`Click [here](index.md)` --- *auto-conversion* ---> `Click [here](index.html)`
+
+<box type="warning">
+
+If you wish to disable the auto-conversion, you may indicate a `no-convert` tag at the end of your link.
```
Comment (ang-zeyu) [inline docs/userGuide/syntax/links.mbdf:44]:
```suggestion
If you wish to disable the auto-conversion, you may use the `no-convert` attribute in your link. 
```

File: docs/userGuide/syntax/links.mbdf [hunk 6]
```diff
@@ -30,6 +30,27 @@ MarkBind home is at [here][1].
...
 Links to files of the generated site (e.g., an HTML page or an image file) can be specified either as relative paths or absolute paths.
 
+****Auto-conversion of extension****
+
+<div class="indented">
+
+Do indicate markdown files with its original extension (**.md**) as it will automatically be converted to a html extension (**.html**) when the site is generated. 
+
+{{ icon_example }}
+`Click [here](index.md)` --- *auto-conversion* ---> `Click [here](index.html)`
```
Comment (ang-zeyu) [inline docs/userGuide/syntax/links.mbdf:40]: nice arrow! 😄

File: docs/userGuide/syntax/links.mbdf [hunk 7]
```diff
@@ -30,6 +30,27 @@ MarkBind home is at [here][1].
 
 Links to files of the generated site (e.g., an HTML page or an image file) can be specified either as relative paths or absolute paths.
 
+****Auto-conversion of extension****
+
+<div class="indented">
+
+Do indicate markdown files with its original extension (**.md**) as it will automatically be converted to a html extension (**.html**) when the site is generated.
```
Comment (ang-zeyu) [inline docs/userGuide/syntax/links.mbdf:37]:
```suggestion
You may link to markdown files using its original extension (**.md**) as it will automatically be converted to a html extension (**.html**) when the site is generated. 
```

**Review by ang-zeyu — APPROVED** [review]

**Comment by ang-zeyu** [comment]
Looks good! : )
Changed the wording just a little
